### PR TITLE
Add simple_test to dense heads.

### PR DESCRIPTION
**Pull request by jshilong**
This PR is based on #5061.
Thanks to @shinya7y for his contribution.

This PR improves the consistency of implementation (single-stage vs. two-stage, normal dense heads vs. RPN heads, simple_test vs. aug_test). Besides it,  it increases the flexibility for  One-Stage Instance segmentation in the future.
Related discussion: https://github.com/open-mmlab/mmdetection/pull/3638#pullrequestreview-493166339

### CODE REVIEW

File: mmdet/models/dense_heads/anchor_head.py [hunk 1]
```diff
@@ -736,6 +736,10 @@ def aug_test(self, feats, img_metas, rescale=False):
                 Defaults to False.
 
         Returns:
-            list[ndarray]: bbox results of each class
+            list[tuple[Tensor, Tensor]]: Each item in result_list is 2-tuple.
+                The first item is bboxes with shape (n, 5), where 5 represent
+                (tl_x, tl_y, br_x, br_y, score).
+                The shape of the second tensor in the tuple is labels
+                with shape (n,), Length of list should always be 1.
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/anchor_head.py:743]: . The length xxx

File: mmdet/models/dense_heads/anchor_head.py [hunk 2]
```diff
@@ -736,6 +736,10 @@ def aug_test(self, feats, img_metas, rescale=False):
                 Defaults to False.
 
         Returns:
-            list[ndarray]: bbox results of each class
+            list[tuple[Tensor, Tensor]]: Each item in result_list is 2-tuple.
+                The first item is bboxes with shape (n, 5), where 5 represent
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/anchor_head.py:740]: is ``bboxes``

File: mmdet/models/dense_heads/anchor_head.py [hunk 3]
```diff
@@ -736,6 +736,10 @@ def aug_test(self, feats, img_metas, rescale=False):
...
+                (tl_x, tl_y, br_x, br_y, score).
+                The shape of the second tensor in the tuple is labels
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/anchor_head.py:742]:
```
is ``labels``
```

File: mmdet/models/dense_heads/dense_test_mixins.py [hunk 4]
```diff
@@ -1,42 +1,42 @@
...
+            feats (tuple[torch.Tensor]): Multi-level features from the
+                upstream network, each is a 4D-tensor.
+            img_metas (list[dict]): List of image information.
+            rescale (bool, optional): Whether to rescale the results.
+                Defaults to False.
 
         Returns:
-            tuple: (bboxes, scores)
+            list[tuple[Tensor, Tensor]]: Each item in result_list is 2-tuple.
+                The first item is bboxes with shape (n, 5), where 5 represent
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/dense_test_mixins.py:28]: see comments above.

File: mmdet/models/dense_heads/dense_test_mixins.py [hunk 5]
```diff
@@ -1,42 +1,42 @@
...
-            scores = torch.cat(aug_scores, dim=0)
-            return bboxes, scores
+        outs = self.forward(feats)
+        results_list = self.get_bboxes(*outs, img_metas, rescale=rescale)
+        return results_list
 
     def aug_test_bboxes(self, feats, img_metas, rescale=False):
-        """Test det bboxes with test time augmentation.
+        """Test det bboxes with test time augmentation, can be applied in
+        DenseHead except for RPNHead and its variants(GuidedRPN etc.)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/dense_test_mixins.py:39]: variants e.g., Guided RPN etc.

File: mmdet/models/dense_heads/dense_test_mixins.py [hunk 6]
```diff
@@ -96,5 +100,96 @@ def aug_test_bboxes(self, feats, img_metas, rescale=False):
...
+                where 5 represent (tl_x, tl_y, br_x, br_y, score).
+        """
+        rpn_outs = self(x)
+        proposal_list = self.get_bboxes(*rpn_outs, img_metas)
+        return proposal_list
+
+    def aug_test_rpn(self, feats, img_metas):
+        """Test with augmentation for only for RPNHead and
+        its variants(GuidedRPN etc.)
+        Args:
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/dense_test_mixins.py:128]: An empty line above.

File: mmdet/models/dense_heads/dense_test_mixins.py [hunk 7]
```diff
@@ -96,5 +100,96 @@ def aug_test_bboxes(self, feats, img_metas, rescale=False):
...
+    def merge_aug_bboxes(self, aug_bboxes, aug_scores, img_metas):
+        """Merge augmented detection bboxes and scores.
+
+        Args:
+            aug_bboxes (list[Tensor]): shape (n, 4*#class)
+            aug_scores (list[Tensor] or None): shape (n, #class)
+            img_shapes (list[Tensor]): shape (3, ).
+
+        Returns:
+            tuple: (bboxes, scores)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/dense_test_mixins.py:179]: tuple[Tensor]:

File: mmdet/models/dense_heads/dense_test_mixins.py [hunk 8]
```diff
@@ -1,42 +1,42 @@
...
-    """Mixin class for test time augmentation of bboxes."""
 
-    def merge_aug_bboxes(self, aug_bboxes, aug_scores, img_metas):
-        """Merge augmented detection bboxes and scores.
+class BBoxTestMixin(object):
+    """Mixin class for testing det bboxes via DenseHead."""
+
+    def simple_test_bboxes(self, feats, img_metas, rescale=False):
+        """Test det bboxes without test-time augmentation, can be applied in
+        DenseHead except for RPNHead and its variants(GuidedRPN etc.)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/dense_test_mixins.py:17]:
```
for ``RPNHead`` and its variants, e.g., ``GARPNHead``, etc.
```

File: mmdet/models/dense_heads/yolact_head.py [hunk 9]
```diff
@@ -924,6 +930,68 @@ def sanitize_coordinates(self, x1, x2, img_size, padding=0, cast=True):
...
+        scale_factors = tuple(meta['scale_factor'] for meta in img_metas)
+        if all(det_bbox.shape[0] == 0 for det_bbox in det_bboxes):
+            segm_results = [[[] for _ in range(self.num_classes)]
+                            for _ in range(num_imgs)]
+        else:
+            # if det_bboxes is rescaled to the original image size, we need to
+            # rescale it back to the testing scale to obtain RoIs.
+            if rescale and not isinstance(scale_factors[0], float):
+                scale_factors = [
+                    torch.from_numpy(scale_factor).to(det_bboxes[0].device)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolact_head.py:974]: what if scale_factors is a list of float?

**Review by ZwwWayne — APPROVED** [review]

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**